File: web/src/lib/protocol.ts
export type Side = "south" | "north";

export type EndReason = "majority" | "must_feed" | "no_progress" | "resign" | "disconnect";

export interface Stores {
  south: number;
  north: number;
}

export interface LastMove {
  by: Side;
  pit: number;
  captured: number;
}

export interface GameState {
  type: "state";
  game_id: string;
  pits: number[];
  stores: Stores;
  to_move: Side;
  ply: number;
  legal_moves: number[];
  last_move: LastMove | null;
}

export interface AgentBrief {
  id: string;
  name: string;
}

export interface GameStarted {
  type: "game_started";
  game_id: string;
  agent: AgentBrief;
  state: GameState;
}

export interface AgentThinking {
  type: "agent_thinking";
  game_id: string;
  since: number;
}

export interface AgentMoveMsg {
  type: "agent_move";
  game_id: string;
  pit: number;
  thought_ms: number;
  extras: Record<string, unknown>;
}

export interface GameOver {
  type: "game_over";
  game_id: string;
  winner: Side | "draw";
  reason: EndReason;
  final_stores: Stores;
}

export interface ErrorMsg {
  type: "error";
  code: string;
  message: string;
}

export type ServerMessage =
  | GameStarted
  | GameState
  | AgentThinking
  | AgentMoveMsg
  | GameOver
  | ErrorMsg
  | { type: "pong"; t: number };

export interface AgentEntry {
  id: string;
  name: string;
  family: string;
  description: string;
  est_elo: number | null;
}

File: src/oware/server/protocol.py
from typing import Any, Literal

from pydantic import BaseModel, Field

Side = Literal["south", "north"]
EndReason = Literal["majority", "must_feed", "no_progress", "resign", "disconnect"]


class ClientNewGame(BaseModel):
  type: Literal["new_game"]
  agent_id: str
  human_plays: Side = "south"
  seed: int | None = None


class ClientMove(BaseModel):
  type: Literal["move"]
  game_id: str
  pit: int = Field(ge=0, le=5)


class ClientResign(BaseModel):
  type: Literal["resign"]
  game_id: str


class ClientPing(BaseModel):
  type: Literal["ping"]
  t: int


ClientMessage = ClientNewGame | ClientMove | ClientResign | ClientPing


class Stores(BaseModel):
  south: int
  north: int


class LastMove(BaseModel):
  by: Side
  pit: int
  captured: int


class GameState(BaseModel):
  type: Literal["state"] = "state"
  game_id: str
  pits: list[int]
  stores: Stores
  to_move: Side
  ply: int
  legal_moves: list[int]
  last_move: LastMove | None


class AgentBrief(BaseModel):
  id: str
  name: str


class GameStarted(BaseModel):
  type: Literal["game_started"] = "game_started"
  game_id: str
  agent: AgentBrief
  state: GameState


class AgentThinking(BaseModel):
  type: Literal["agent_thinking"] = "agent_thinking"
  game_id: str
  since: int


class AgentMove(BaseModel):
  type: Literal["agent_move"] = "agent_move"
  game_id: str
  pit: int
  thought_ms: int
  extras: dict[str, Any] = Field(default_factory=dict)


class GameOver(BaseModel):
  type: Literal["game_over"] = "game_over"
  game_id: str
  winner: Literal["south", "north", "draw"]
  reason: EndReason
  final_stores: Stores
  history: list[dict] = Field(default_factory=list)


class GameAnalysis(BaseModel):
  type: Literal["game_analysis"] = "game_analysis"
  game_id: str
  history: list[dict]


class ErrorMessage(BaseModel):
  type: Literal["error"] = "error"
  code: str
  message: str


class PongMessage(BaseModel):
  type: Literal["pong"] = "pong"
  t: int
